src/Integrators.jl

```diff
@@ -107,7 +110,9 @@ function step!(sys::System{0}, integrator::ImplicitMidpoint)
         @. s̄′ = s + 0.5 * Δt * rhs_dipole(ŝ, B)
 
         # If converged, then we can return
-        if isapprox(s̄, s̄′, atol=atol* √length(s̄))
+		@. progress = norm(s̄ - s̄′)
+        if maximum(progress) <= atol
```

Comment:
First, it is surprising to learn that `isapprox` is allocating -- in principle, this could be implemented in a non-allocating way. The semantics of the old convergence condition is (I think):
```julia
@. progress = norm(s̄ - s̄′)^2
cond = sum(progress .^2) <= atol* √length(s̄)
```
The new condition uses a maximum instead, which might or might not be what we want. 

I wonder if it's better to have a custom, non-allocating implementation of isapprox, and then use it uniformly in this file? Then the code would be easy to read and we wouldn't need the extra scalar buffers.

Reply:
The difference is L2 vs Linfinity norm. Conceptually, the L2 version is like a soft-spin constraint and Linfinity is like a hard-spin constraint---it's a real physics question about which is a better stopping criterion for the method.

I think custom `isapprox` is a very good plan. Ideally, it doesn't even need to check every spin and probably could stop after the first one in the most common case.

Reply:
> I think custom `isapprox` is a very good plan. Ideally, it doesn't even need to check every spin and probably could stop after the first one in the most common case.

Yes, the possibility of "early return" is an advantage of Linfinity, and I agree that it's not clear if there's a strong physics reason to prefer one norm over the other. If we define a custom `isapprox` then it would be easy to play with it later.